### ISSUE TITLE
Amplify configure throwing NotAuthorizedException when refresh token has expired

### ISSUE THREAD

**Issue by 94Sip** [aws-amplify/amplify-flutter#3950]
### Description

I am receiving a underlyingexception of NotAuthorizedExcpetion when I call Amplify.configure().

My questions are this:

1. Shouldn't I be able to configure Amplify regardless of the refresh token?
2. the recovery suggestion is to call SignIn, but I don't understand how to do that here since I have not gone thru Authenticator yet? Does that matter? 
3. If I call fetchAuthSession, it requires the Auth plugin, which hasn't been configured yet, right?

Below is a screenshot of the error I am catching, and also, here is the main.dart file which configures Amplify.
```dart
import 'dart:async';
import 'dart:ui';
import 'package:amplify_push_notifications_pinpoint/amplify_push_notifications_pinpoint.dart';
import 'package:beamer/beamer.dart';

// Firebase packages and config
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/scheduler.dart';
import './custom_error.dart';
import '../../ranch/view/ranch_info.dart';
import '../../util/progress_widget.dart';
import 'firebase_options.dart';
import 'package:firebase_crashlytics/firebase_crashlytics.dart';
//import 'package:firebase_messaging/firebase_messaging.dart';
//import '../util/notifications.dart';

import 'package:flutter/material.dart';
import 'package:flutter_bloc/flutter_bloc.dart';
import 'package:authentication_repository/authentication_repository.dart';
import '../../analysis/analyze.dart';
import '../../locations/bloc/locations_repository.dart';
import '../../appfunctions/appfunctions.dart';
import 'configure_nonweb.dart' if (dart.library.html) 'configure_web.dart';

// Amplify Flutter Packages
import 'package:amplify_flutter/amplify_flutter.dart';
import 'package:amplify_api/amplify_api.dart';
import 'package:amplify_storage_s3/amplify_storage_s3.dart';
import 'package:amplify_auth_cognito/amplify_auth_cognito.dart';
import 'package:amplify_authenticator/amplify_authenticator.dart';
import 'package:amplify_analytics_pinpoint/amplify_analytics_pinpoint.dart';
import 'amplifyconfiguration.dart';
import 'ampconfig_nopn.dart';

import 'home/home.dart';
import '../themes/colors.dart';

//import 'package:flutter/cupertino.dart';
import '../../logging.dart';
import 'platform_specific.dart';
import 'review/bloc/review_bloc.dart';

// Note: This handler does not *need* to be async, but it can be!
//this is placed here so it can be invoked even when app is terminated

Future<void> backgroundAsyncNotificationReceivedHandler(
    PushNotificationMessage notification) async {
  final count = await Amplify.Notifications.Push.getBadgeCount();
  Amplify.Notifications.Push.setBadgeCount(count + 1);
}

/// This defines if we are using the iOS simulator or not for testing
const simulating = bool.hasEnvironment("SIMULATOR");

void main() async {
  if (const bool.hasEnvironment("DEBUG")) {
    AmplifyLogger().logLevel = LogLevel.verbose;
  } else {
    AmplifyLogger().logLevel = LogLevel.error;
  }

  WidgetsFlutterBinding.ensureInitialized();

  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );
  bool weWantFatalErrorRecording = true;
  late AmplifyPushNotificationsPinpoint notificationsPlugin;

  if (!simulating) {
    notificationsPlugin = AmplifyPushNotificationsPinpoint();

    // Should be added in the main function to avoid missing events and before Amplify.configure()
    notificationsPlugin.onNotificationReceivedInBackground(
        backgroundAsyncNotificationReceivedHandler);
  }

  // this handles all errors caught by Flutter
  //this is typically called just prior to ErrorWidget.builder being invoked
  FlutterError.onError = (errorDetails) {
    if (weWantFatalErrorRecording) {
      FirebaseCrashlytics.instance.recordFlutterFatalError(errorDetails);
    } else {
      FirebaseCrashlytics.instance.recordFlutterError(errorDetails);
    }
  };
  // Pass all uncaught asynchronous errors that aren't handled by the Flutter framework to Crashlytics
  PlatformDispatcher.instance.onError = (error, stack) {
    //FirebaseCrashlytics.instance.recordError(error, stack, fatal: true);
    return true;
  };

  //NOTE: this handles errors that occur during the build process
  ErrorWidget.builder = (FlutterErrorDetails errorDetails) {
    return CustomError(errorDetails: errorDetails);
  };

  configureApp();
  if (simulating) {
    runApp(const SmartHuntApp());
  } else {
    runApp(SmartHuntApp(
      pinPointPNplugin: notificationsPlugin,
    ));
  }
}

class SmartHuntApp extends StatefulWidget {
  const SmartHuntApp({Key? key, this.pinPointPNplugin}) : super(key: key);

  final AmplifyPushNotificationsPinpoint? pinPointPNplugin;

  @override
  AppViewState createState() => AppViewState();
}

class AppViewState extends State<SmartHuntApp> {
  bool _isAmplifyConfigured = false;
  final log = logger(SmartHuntApp);
  final platform = PlatformSpecific(switchUI: false);
  bool isLoggedIn = false;
  late final AppLifecycleListener _listener;
  late AppLifecycleState? _state;
  late StreamSubscription foreGroundSubscription,
      notificationOpenedSubscription,
      tokenSubscription;

  @override
  void initState() {
    super.initState();
    _configureAmplify();
    _state = SchedulerBinding.instance.lifecycleState;
  }

  @override
  void dispose() {
    _listener.dispose();
    if (simulating) {
      foreGroundSubscription.cancel();
      notificationOpenedSubscription.cancel();
      tokenSubscription.cancel();
    }
    super.dispose();
  }

  void _handleStateChange(AppLifecycleState state) {
    setState(() {
      _state = state;

      log.i("state changed: $_state");
    });
  }

  @override
  Widget build(BuildContext context) {
    return _isAmplifyConfigured
        ? MultiRepositoryProvider(
            providers: [
              RepositoryProvider(
                create: (context) => AuthenticationRepository(),
                lazy: false,
              ),
              RepositoryProvider(create: (context) => LocationsRepository())
            ],
            child: MultiBlocProvider(providers: [
              BlocProvider<AppfunctionsBloc>(
                  lazy: false,
                  create: (context) => AppfunctionsBloc(
                      locRepo: context.read<LocationsRepository>(),
                      authenticationRepository:
                          context.read<AuthenticationRepository>())),
              BlocProvider<AnalyzeBloc>(
                  create: (context) => AnalyzeBloc(
                      locRepo: context.read<LocationsRepository>())),
              BlocProvider<ReviewBloc>(
                  create: (context) =>
                      ReviewBloc(locRepo: context.read<LocationsRepository>())),
            ], child: CustomAuthenticator(routerDelegate: routerDelegate)),
          )
        : MaterialApp(
            home: const ProgressWidget(
              whatIsThisDoing: "Starting up",
            ),
            theme: MyTheme.defaultTheme,
          );
  }
  //TODO: figure out how to NOT route to create ranch screen and then have to route to home screen
  //if they are logged in and have a ranchid

  //this is the app's top-level router - it simply points to ScaffoldWithBottomNavBar, which then has multiple nested routers
  // This also has a guard that prevents the user from accessing any of the app functions if the ranchid has not been set
  final routerDelegate = BeamerDelegate(
    initialPath: '/home/cameras',
    routeListener: (routeInformation, delegate) {
      //this is called on every navigation change
      //so it is called before the guards are applied since the initialPath is set
      final location = routeInformation.uri.path;
    },
    buildListener: (context, delegate) {
      final location = Beamer.of(context).configuration.uri.path;
    },
    locationBuilder: RoutesLocationBuilder(
      routes: {
        '/home/*': (context, state, data) {
          return const ScaffoldWithBottomNavBar();
        },
        "/ranch": (context, state, data) {
          context.read<AppfunctionsBloc>().add(const CheckForRanch());

          return const RanchScreen();
        },
      },
    ),
    guards: [
      // Guard all paths not matching /ranch if the ranchid has not been set
      BeamGuard(
        pathPatterns: ['/ranch'],
        guardNonMatching: true,
        check: (context, location) {
          if (context.read<AppfunctionsBloc>().state.user!.ranchId.isEmpty) {
            return false;
          } else {
            return true;
          }
        },
        beamToNamed: (_, __) => '/ranch',
      ),
    ],
  );

  Future<void> _configureAmplify() async {
    if (Amplify.isConfigured == false) {
      try {
        await Amplify.addPlugins([
          AmplifyAuthCognito(),
          AmplifyAPI(),
          AmplifyStorageS3(),
          if (!simulating) widget.pinPointPNplugin!,
          AmplifyAnalyticsPinpoint()
        ]);
        log.i("Amplify plugins added");
        //amplifyconfig defined in file amplifyconfiguration.dart which is imported above
        if (simulating) {
          await Amplify.configure(amplifyconfigNoPN);
        } else {
          await Amplify.configure(amplifyconfig);
        }

        log.i("Amplify configured successfully");

        // Required to call this after Amplify.configure.
        // Doesn't get called on app start as event is swallowed by library to register device.
        if (!simulating) {
          tokenSubscription =
              Amplify.Notifications.Push.onTokenReceived.listen((event) {
            log.i('onTokenReceived $event');
          });

          // Required to call this after Amplify.configure.
          //Notifications received while your app is in the foreground state do not get displayed in notification center
          foreGroundSubscription = Amplify
              .Notifications.Push.onNotificationReceivedInForeground
              .listen((event) {
            log.i('onNotificationReceivedInForeground $event');
          }, onError: (e) {
            log.e("onNotificationReceivedInForeground error: $e");
          },
                  onDone: () =>
                      log.i("onNotificationReceivedInForeground done"),
                  cancelOnError: false);

          // Required to call this after Amplify.configure.
          //this is called when the user taps on the notification in notification center and your app is in foreground
          notificationOpenedSubscription =
              Amplify.Notifications.Push.onNotificationOpened.listen((event) {
            log.i('onNotificationOpened $event');
          }, onError: (e) {
            log.e("onNotificationOpened error: $e");
          },
                  onDone: () => log.i("onNotificationOpened done"),
                  cancelOnError: false);
        }
        // Required to call this after Amplify.configure.
        _listener = AppLifecycleListener(
          // This fires for each state change.
          onStateChange: _handleStateChange,
        );
        setState(() => _isAmplifyConfigured = true);

        if (_state != null) {
          log.i("initial state:  $_state");
        }
      } catch (e) {
        log.e("Error configuring Amplify: {$e}");
        //TODO: handle expired token
      }
    } else {
      log.i("Amplify already configured");
    }
  }
}

class CustomAuthenticator extends StatelessWidget {
  CustomAuthenticator({
    super.key,
    required this.routerDelegate,
  });

  final BeamerDelegate routerDelegate;
  final log = logger(CustomAuthenticator);

  @override
  Widget build(BuildContext context) {
    return BlocListener<AppfunctionsBloc, AppfunctionsState>(
        listenWhen: (previous, current) {
          return current.appFunction == AppFunctions.login;
        },
        listener: (context, state) {
          log.i("updating routerDelegate due to login appfunction");
          //routerDelegate.update();
        },
        child: Authenticator(
          onException: (exception) {
            log.e(exception);
          },
          child: MaterialApp.router(
            theme: MyTheme.defaultTheme,
            //using the Authenticator.builder() ensures that all routes are protected by authentication
            builder: Authenticator.builder(),
            routerDelegate: routerDelegate,
            routeInformationParser: BeamerParser(),
            backButtonDispatcher: BeamerBackButtonDispatcher(
              delegate: routerDelegate,
            ),
          ),
        ));
  }
}
```

### Categories

- [ ] Analytics
- [ ] API (REST)
- [ ] API (GraphQL)
- [X] Auth
- [X] Authenticator
- [ ] DataStore
- [ ] Notifications (Push)
- [ ] Storage

### Screenshots

<img width="964" alt="amplifyconfigerror" src="https://github.com/aws-amplify/amplify-flutter/assets/60077534/61acf302-0b22-4aea-bb77-45d82ee4142f">

The above is the screenshot of the error I am catching during Amplify.configure().

### Platforms

- [X] iOS
- [X] Android
- [ ] Web
- [ ] macOS
- [ ] Windows
- [ ] Linux

### Flutter Version

3.13.5

### Amplify Flutter Version

1.4.x

### Deployment Method

Amplify CLI

**Comment by haverchuck** [comment]
@94Sip Thanks for submitting this - we are working on a resolution.

**Comment by Jordan-Nelson** [comment]
@94Sip this has been resolved in v1.5.0. Thanks for your patience while we got a fix out.
